docs/extensibility/custom-resources.md

```diff
@@ -0,0 +1,432 @@
+    message.Subject = "You are unsubscribed from our newsletter!";
+    message.Body = "Sorry to see you go. We hope you will come back soon!";
+    await smtpClient.SendMailAsync(message);
+});
+```
+
+Once the `Program.cs` file is updated launch the AppHost and use your browser, or `curl` to hit the following URLs:
+
+```bash
+# Remember to replace the port to match your local environment.
+curl https://localhost:7289/subscribe?email=test@test.com
+curl https://localhost:7289/unsubscribe?email=test@test.com
+```
+
+If those API calls return a successful response then you should be able to click on the `maildev` resource the dashboard and the MailDev UI will show the emails that have been sent to the SMTP endpoint.
+
+![E-mails visible in the MailDev UI](./media/maildev-emails.png)
+
+## Technical details
+
+### IValueProvider and IManifestExpressionProvider
+
+In the code above the `MailDevResource` had two properties. `SmtpEndpoint` and `ConnectionStringExpression`. The types of these properties were <xref:Aspire.Hosting.ApplicationModel.EndpointReference> and <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> respectively. These types are among several which are used throughout .NET Aspire to represent configuration data which is not finalized until the .NET Aspire application is either run or published to the cloud via a tool such as `azd` (Azure Developer CLI).
+
+The fundamental problem that these types help solve is deferring resolution of concrete configuration information until _all_ of the information is available.
+
+For example, in the `MailDevResource` example we expose a property called `ConnectionStringExpression` as required by the <xref:Aspire.Hosting.ApplicationModel.IResourceWithConnectionString> interface. The type of the property is <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> and is created by passing in an interpolated string to the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create> method.
+
+```csharp
+public ReferenceExpression ConnectionStringExpression => ReferenceExpression.Create(
+        $"smtp://{SmtpEndpoint.Property(EndpointProperty.Host)}:{SmtpEndpoint.Property(EndpointProperty.Port)}"
+        );
+```
+
+The signature for the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create> method is as follows:
+
+```csharp
+public static ReferenceExpression Create(in ExpressionInterpolatedStringHandler handler)
+```
+
+Notice that this is not a regular <xref:System.String> argument. Method makes use of the [interpolated string handler pattern](../../csharp/whats-new/tutorials/interpolated-string-handler) in .NET to capture the interpolated string template and the values referenced within it to allow for custom processing. In the case of .NET Aspire we capture these details in a <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> which can be evaluated as each value referenced in the interpolated string becomes available.
```

Comment:
The reason that this `csharp` doesn't work, is that you're trying to use relative path resolution on a file that's not found in this docset. Instead, you'd want to use site relative path resolution. This has all the details, if you're interested in learning about how it's done: https://review.learn.microsoft.com/en-us/help/platform/links-how-to?branch=main

```suggestion
Notice that this is not a regular <xref:System.String> argument. Method makes use of the [interpolated string handler pattern](/dotnet/csharp/whats-new/tutorials/interpolated-string-handler) in .NET to capture the interpolated string template and the values referenced within it to allow for custom processing. In the case of .NET Aspire we capture these details in a <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> which can be evaluated as each value referenced in the interpolated string becomes available.
```